Introduce restricted journals for work package comments with restricted visibility

akabiru opened the pull request:
# Ticket

- [Database migration changes](https://community.openproject.org/work_packages/60980)
- [Restricted Comments UI](https://community.openproject.org/wp/60979)
- [Add feature flag](https://community.openproject.org/projects/communicator-stream/work_packages/62009/activity)
- [Add new permissions for restricted work package comments](https://community.openproject.org/projects/communicator-stream/work_packages/60978/activity)
- [Honor work package comment with restricted visibility permissions on the activity tab (in both notification center and wp full view)](https://community.openproject.org/projects/communicator-stream/work_packages/60989/activity)

# What are you trying to accomplish?

This is a partial implementation of comments with restricted visibility. It is included here data and query changes on journals, so that we can hide/show restricted ones based on visibility rules.

https://github.com/user-attachments/assets/629aac33-c0a9-4e48-98a5-8315ef379786

# For reviewers

The code in this PR was mostly reviewed as other partial PRs that were merged in this branch. We would still very much appreciate a sanity check on database migrations and journal changes.

akabiru committed:
Introduce work package restricted comments as part of journal notes

https://community.openproject.org/work_packages/60980

brunopagno committed:
add new permissions for comments with restricted visibility

akabiru committed:
Bare-bones for writing restricted journals

TODO:
 1) Account for associated journals such as attachments, ensure those are restricted
 2) Exclude restricted journals from aggregation? (Unclear)
 3) Exclude comment-only journals from "snapshots" i.e. making a copy of the current wp state

akabiru committed:
Exclude journals with restricted flag from aggregation

Atm, the "restricted" concept only exists for comments - aka "Journal#notes"
hence, at the momement, there other changes should not be aggregated to a restricted
journal.

brunopagno committed:
add default permissions

akabiru committed:
Update work package note contract with restriction
Do not allow empty notes with restriction, as empty notes are not seen as comments

akabiru committed:
Add unit test case for restricted note in service

akabiru committed:
[#60979] Add capability to create work package comments with restricted visibility (UI)

https://community.openproject.org/work_packages/60979

akabiru committed:
[#60979] Add capability to create work package comments with restricted visibility (UI)

https://community.openproject.org/work_packages/60979

brunopagno commented on app/contracts/work_packages/create_note_contract.rb:
```diff
@@ -28,14 +30,16 @@
       errors.add(:journal_notes, :error_unauthorized) unless can?(:comment)
     end
 
+    attribute :journal_restricted do
+      errors.add(:journal_restricted, :invalid) if model.journal_notes.blank? && model.journal_restricted
+    end
```
I think it makes things a bit simpler if we don't make a special case for notes. This check seems correct from the point of view of the "comments with restricted visibility" feature, but not really from journals as a db model

akabiru replied:
Keeping in mind this contract is a `CreateNoteContract` which is only used in the `AddWorkPackageNoteService` specifically for wp journal notes and not in any other journal creation service it would seem fitting. Pls lmk if I'm missing some context

brunopagno replied:
I see. That makes much more sense now. :thinking:
All good, we can mark it as resolved :bow:

brunopagno commented on app/services/journals/create_service.rb:
```diff
@@ -901,6 +904,10 @@ def only_one_note(predecessor, notes)
       predecessor.notes.empty? || notes.empty?
     end
 
+    def unrestricted_journal?(restricted)
+      !restricted
+    end
```
I think it would still be better to do this check as

```ruby
  predecessor.restricted == restricted
```

My thought is that if the previous and current updates are both restricted we should be okay aggregating them.

brunopagno reviewed (commented):
I'll say this looks absolutely briliant. Thanks a lot for digging into the journal behaviour :bow: 

From my side, I think we are okay pushing this forward, but I did leave a few comments with ideas that _in my insane mind_ might make things simpler for us in the future.

That said, I think it would probably worth getting the review of someone with a bit more mileage in this part of the code, just for sanity check of our assumptions.

brunopagno committed:
hide restricted journals when user does not have access

akabiru committed:
Mark restricted comment with different color shades

The color hex codes are bound to change as we should default to primer colors
that work in all color modes

akabiru committed:
change primer icon size so it fits well with lock icon

https://community.openproject.org/wp/61941

brunopagno committed:
Merge pull request #18095 from opf/impl/60978-new-permissions-for-comments-with-restricted-visibility

Add new permissions for comments with restricted visibility

akabiru committed:
Draft out restricted comment box

brunopagno committed:
include specs

akabiru committed:
Implement design refresh with primer colors

Design [OPEN POINT](https://community.openproject.org/wp/61949)

akabiru committed:
Relocate restricted comment stimulus controller closer to comment

akabiru committed:
Add visibility explainer

akabiru committed:
Use simpler default caption interface

1.) Use primer default (form) layout
2.) Better for a11ly as there's no dissapearing and reappearing text
3.) Consistent on mobile layouts as no wrapping is done on smaller widths

akabiru committed:
Use domain names for stimulus action and target

akabiru committed:
Use cascade instead of primer classes

akabiru committed:
Adapt existing BEM format

akabiru committed:
Revert "Use simpler default caption interface"

This reverts commit be52e1f5bde99aee39df6ecede9f8b46258219f1.

akabiru committed:
Add border top to restricted comment input

akabiru committed:
Revert "Revert "Use simpler default caption interface""

This reverts commit fd39fb7279987f308f44e5b203a6b2e0fcba4417.

brunopagno committed:
implement scope for fetching restricted journals

akabiru committed:
fix: ensure form state is reset on submit

akabiru committed:
Ensure background color is reset on form submit end

akabiru committed:
Merge branch 'dev' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

brunopagno committed:
Update spec/models/journal_spec.rb

Co-authored-by: Kabiru Mwenja <k.mwenja@openproject.com>

akabiru committed:
Use same restriction aggregation model

Further adopt the notion of restricted journals.

Co-authored-by: Bruno Pagno <b.pagno@openproject.com>

akabiru committed:
Merge branch 'dev' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

akabiru committed:
Merge remote-tracking branch 'origin/impl/60978-new-permissions-for-comments-with-restricted-visibility' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

akabiru committed:
Merge branch 'implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes' into impl/60989-hide-restricted-comments

brunopagno committed:
Merge pull request #18131 from opf/impl/60989-hide-restricted-comments

Reject restricted journals when user cannot see them

brunopagno committed:
introduce feature flag for comments with restricted visibility

akabiru committed:
Declare restricted comment highlight class as stimulus class attributes

> Instead of hard-coding classes with JavaScript strings, Stimulus lets you refer to CSS classes by logical name using a combination of data attributes and controller properties.
> See: https://stimulus.hotwired.dev/reference/css-classes

akabiru committed:
Simplify highlight toggle via `classList.toggle()`

> https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle

brunopagno committed:
Merge pull request #18217 from opf/impl/62009-restricted-comments-feature-flag

Feature flag for comments with restricted visibility

brunopagno commented:
Hey @ulferts since you're the go to reference on journals we would apprciate your sanity check on this PR. The description should cover the current state of things, but let us know if anything.

akabiru committed:
Adopt permission model for restricted comments

akabiru committed:
Include test scenarios for restricted comments edit

akabiru committed:
Revert "Simplify highlight toggle via `classList.toggle()`"

This reverts commit 3213e2561d51ba9d95412c2410eeff0b12a75b05.

The input checked status is actually needed

akabiru commented on config/initializers/permissions.rb:
```diff
@@ -309,6 +309,34 @@
+                     dependencies: :view_work_packages,
+                     visible: -> { OpenProject::FeatureDecisions.comments_with_restricted_visibility_active? }
+
+      wpt.permission :write_comments_with_restricted_visibility,
```
🟢 _nitpick_ @brunopagno I wonder if `add_comments_with_restricted_visibility` is more fitting and conventional to OP, the term write can imply both add + edit actions which can be confusing

brunopagno replied:
Good point. The name "write" came from product and I did not think about this. The actual _action_ would be `add`. Lets remember to adjust this before merging, then.

brunopagno replied:
https://github.com/opf/openproject/pull/18253

brunopagno committed:
rename permission from write to add

akabiru committed:
Remove leftover debug logs

brunopagno committed:
Merge pull request #18253 from opf/impl/60978-rename-permission

rename permission to add_comments_with_restricted_visibility

akabiru committed:
Extend comments with restricted visibility feature flag (#18245)

akabiru committed:
Update add restricted comment permission name

akabiru committed:
feature flag icon size

See https://community.openproject.org/wp/61941 for final decision

akabiru committed:
Merge pull request #18144 from opf/implementation/60979-add-capability-to-create-work-package-comments-with-restricted-visibility-(ui)

Implementation/60979 add capability to create work package comments with restricted visibility (UI)

akabiru committed:
Amend editor dismiss strategy to be less eager

Once editor is open, allow for other form interactions such as the restricted visibility checkbox.
User can dismiss the editor via `Esc` key- if there's no content. Else- there will be a cancel button to
explicitly close out the editor

akabiru committed:
Delete unused function

akabiru committed:
Use safe navigation syntax sugar

akabiru committed:
Add feature specs for Esc key dismiss strategy

akabiru committed:
Require confirmation on Kbd Esc with editor content

akabiru committed:
Enable comments with restricted visibility feature flag in pull preview

akabiru committed:
Merge branch 'dev' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

akabiru committed:
[rubocop] Use `expect(journal: [:notes, :restricted])` instead.

See https://github.com/rails/rails/pull/51674 for context.
OR https://gorails.com/episodes/how-to-use-params-expect-rails-8

akabiru committed:
Touch up aggregation explainer

akabiru committed:
Merge branch 'implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes' into implementation/62138-adjust-activity-tab-comment-dismiss-strategy-to-include-cancel-button-and-esc-keyboard-action

ulferts commented on app/contracts/work_packages/create_note_contract.rb:
```diff
@@ -28,14 +30,16 @@
     end
 
+    attribute :journal_restricted do
+      errors.add(:journal_restricted, :invalid) if model.journal_notes.blank? && model.journal_restricted
```
Didn't look into this too deeply. But would it make sense to check for the feature flag in here as well?

akabiru replied:
I wasn't sure before- but it doesn't hurt to implement additional safety. I reckon it would mean that when the feature flag is off- and the user hacks in sending `retsricted=true`- the contract fails validation

ulferts commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -295,13 +295,14 @@ def update_index_component
-            send_notifications: !(params.has_key?(:notify) && params[:notify] == "false"))
-    ###
+            send_notifications: to_boolean.call(params[:notify], true),
+            restricted: to_boolean.call(journal_params[:restricted], false))
```
Since the intend of the to_boolean proc doesn't seem to pass it around, it would be easier to just extract the functionality into a simple method of this controller.

akabiru replied:
Makes sense- I was trying a new "localised" approach. Indeed a method is simpler. 👍🏾

ulferts commented on app/models/journal.rb:
```diff
@@ -113,6 +113,14 @@ class Journal < ApplicationRecord
+      all
+    else
+      where(restricted: false)
+    end
```
This I fear is prone to producing errors. At least when used in a way that is not intended by this PR. 

Within this PR, the calls are like this:
```
work_package
  .journals
  .restricted_visible(work_package
```

Doing so will return the visible journals of this work package.

But nothing prevents somebody from calling

```
Journal
  .restricted_visible(work_package)
```

which, depending on the permissions, might translate to

```
Journal.all
```

And that, after the permission has been checked on a single work package.

This seems to be a prime use case for an [extension](https://guides.rubyonrails.org/association_basics.html#extensions) on the `journals` association.

That way, the scope would only be visible on the `journals` association and it would read nicer as the caller wouldn't have to provide the `work_package` again:

```
work_package
  .journals
  .restricted_visible
```

akabiru replied:
Oh wow, I see how this can be misused. The extensions idea is sound. We'll look into that! 👍🏾

akabiru replied:
@brunopagno looking into this one 👀

akabiru replied:
https://github.com/opf/openproject/pull/18366

ulferts commented on config/initializers/permissions.rb:
```diff
@@ -309,6 +309,34 @@
+                     {},
+                     permissible_on: %i[work_package project],
+                     require: :loggedin,
+                     dependencies: %i[view_work_packages view_comments_with_restricted_visibility],
```
Is the intend to be able to set the permissions on individual work packages? So far, I haven't seen this manifested in the PR, at least in the UI. I see also calls for `User.current.allowed_in_work_package?` which suggests that this is intended. But it seems to me that the permissions are currently only seeded for project permissions. And since there is no UI currently for the work package roles, users will not be able to assign the permission themselves. 

I could see valid use cases for doing this but the reason why I am asking is performance. Querying for work package permissions is way more costly than querying on the project level.

brunopagno replied:
Good point. I believe the product requirement is only project permissions, not permission for individual work packages.

We can make the change.

brunopagno replied:
This was adjusted and merged into the feature branch. Resolving the comment

ulferts commented on app/models/activities/fetcher.rb:
```diff
@@ -114,6 +115,16 @@ def eager_load_associations(events)
+        else
+          event.journal.restricted?
+        end
+      end
```
This here contradicts what is written in the `restricted_visible` scope where the check is on the work package level as well. 

Additionally, it will be slow without benefits I can see. Adapting the query for fetching the journals in the first place is the way to go, I think.

To this end, a subquery can be added when fetching the journals. That could either be

```
WorkPackage.allowed_to(User.current, :view_comments_with_restricted_visibility)
```

or

```
Project.allowed_to(User.current, :view_comments_with_restricted_visibility)
```

depending on whether the permission can or cannot be assigned to work package roles as well.

brunopagno replied:
@akabiru fyi, I'll look into this one, next

brunopagno replied:
The fix https://github.com/opf/openproject/pull/18367

ulferts commented on app/components/work_packages/activities_tab/journals/new_component.html.erb:
```diff
@@ -53,28 +53,41 @@
+            concat(
+              flex_layout do |flex|
+                flex.with_row(mb: 3, test_selector: "op-work-package-journal-restricted-comment-checkbox") do
+                  render(WorkPackages::ActivitiesTab::Journals::RestrictedNoteForm.new(f))
```
Clicking on the checkbox closes the comment field unless the user inserts text first. So if the user has the flow of first selecting the visibility, this is annoying.

![Kapture 2025-03-13 at 10 30 36](https://github.com/user-attachments/assets/c7bde95f-8efc-4bef-b05c-310cf4ef0f95)

brunopagno replied:
This is a known issue. We are making design changes to the whole comment compose input as follow ups.

akabiru replied:
Yes- covered in #18262, #18302

ulferts commented on app/models/work_package.rb:
```diff
@@ -382,8 +382,18 @@ def update_done_ratio_from_status
+    can_edit_journal_note = ->(edit_others_notes_permission, edit_own_notes_permission) {
+      user.allowed_in_project?(edit_others_notes_permission, project) ||
+        (user.allowed_in_work_package?(edit_own_notes_permission, self) && journal.user_id == user.id)
+    }
```
This I would also extract into a simple method that is then called.

ulferts replied:
This also has both calls in it. The one on the project and the one on the work package level.

ulferts reviewed (changes requested):
I haven't completed my review yet. From what I have seen though, there are a number of contradictions and things to improve but the PR is heading into a direction that this can be merged. 

Personally, I am still convinced that the commenting functionality should be extracted from journalizing but within the constraints of excluding this possibility, this PR works.

Appart from the comments within the code, what I would see as a prerequisite for merging, are tests added on the `acts_as_journalized_level`, e.g. for [work packages](https://github.com/opf/openproject/blob/dev/spec/models/work_package/work_package_acts_as_journalized_spec.rb)

ulferts commented on app/services/journals/create_service.rb:
```diff
@@ -62,24 +62,28 @@ def call(notes: "", cause: CauseOfChange::NoCause.new)
+    # * The journalizing happens within the configured aggregation time
+    # * The journalizing is carried out by the same user
+    # * The cause is identical
+    # * ONLY the predecessor or the journal to be created has notes
```
Aware that this was taken from the comment before, I would still rewrite this as:

'ONLY one note exists between the predecessor and the journal to be created'

akabiru replied:
Ah, yes- typo! 👍🏾

ulferts commented on app/services/journals/create_service.rb:
```diff
@@ -62,24 +62,28 @@ def call(notes: "", cause: CauseOfChange::NoCause.new)
+    # * The journalizing is carried out by the same user
+    # * The cause is identical
+    # * ONLY the predecessor or the journal to be created has notes
+    # * The predecessor and the journal to be created have the restriction
```
This sentence seems to be incomplete. At least as of now, that would mean that either both are restricted or both are not.

In the broader context, I don't know if that actually needs to be checked as, at least currently, a restricted journal will always have a note. But I guess it is better to check this explicitly in case the requirements change.

ulferts commented on app/components/work_packages/activities_tab/index_component.rb:
```diff
@@ -80,6 +93,8 @@ def polling_interval
         User.current.allowed_in_work_package?(:add_work_package_notes, @work_package)
       end
+
+      def activities_tab_index_stimulus_controller = "work-packages--activities-tab--index"
```
Why extract this to a method and not the call in [`add_comment_wrapper_data_attributes`](https://github.com/opf/openproject/pull/18091/files#diff-51c7ddab80242dd6f1c64cf7c9f8fc05e7d239ffc0e53eef728117caf1f9137bR72)

Personally, I would also prefer having it extracted to a method one can then just call where the suffix is appended. I think it would read nicer, e.g.:

```
          index_stimulus_controller("-update-streams-path-value") => :update_streams_work_package_activities_path(work_package),
          index_stimulus_controller("-sorting-value") => journal_sorting,
          index_stimulus_controller("-filter-value") => filter,
          index_stimulus_controller("-user-id-value" => User.current.id,
          index_stimulus_controller("-work-package-id-value") => work_package.id,
          index_stimulus_controller("-polling-interval-in-ms-value") => polling_interval,
          index_stimulus_controller("-notification-center-path-name-value") => notifications_path,
          index_stimulus_controller("-show-conflict-flash-message-url-value") => show_conflict_flash_message_work_packages_path,
```

But admittedly, that is subjective, so feel free to ignore my comment.

akabiru replied:
It's actually called in the [`action: "#{activities_tab_index_stimulus_controller}:onSubmit-end@window->#{stimulus_controller}#onSubmitEnd",`](https://github.com/opf/openproject/pull/18091/files#diff-51c7ddab80242dd6f1c64cf7c9f8fc05e7d239ffc0e53eef728117caf1f9137bR79) that passes events between the index and comments controller.

Good call, the proposed structure with suffix does read nicer. Will update! 👍🏾

ulferts commented on app/components/work_packages/activities_tab/journals/new_component.rb:
```diff
@@ -57,6 +57,11 @@ def button_row_display_value
+
+        def adding_restricted_comment_allowed?
+          OpenProject::FeatureDecisions.comments_with_restricted_visibility_active? &&
+            User.current.allowed_in_work_package?(:add_comments_with_restricted_visibility, work_package)
```
This line of code again checks for the permission on work package level

ulferts reviewed (changes requested):
I finished my review. To be clear, I didn't do a complete review but rather focused on the backend changes and the structure therein.

It was surprising to see how little changes were necessary on the `Journals::CreateService`. That is good news insofar as that there should be less possibilities for errors.

akabiru committed:
[#62139] Implement redesign of "restricted visibility" checkbox placement in desktop and mobile

https://community.openproject.org/work_packages/62139

akabiru committed:
Rename editor close action to be more logical

akabiru committed:
Attempt to align learn more link

Use padding for now- see if there's a better way

akabiru committed:
Ensure form is reset on cancel

akabiru committed:
Add specs for cancel action + several fixes

akabiru committed:
Let the editor take the full width

akabiru committed:
[rubocop] Consider using explicit block argument in the surrounding method's signature over `yield`.

akabiru committed:
Exclude learn more link on mobile layout
Introduce a base class to reuse mobile layout inference

akabiru committed:
[stimulus] Duplicate definition of Stimulus Value "notificationCenterPathName".

A parent controller already defines this Value.Stimulus LSP (stimulus.source_file.error).
--
_TIL_ In Stimulus, to properly inherit and extend static values without duplicates, you should use the static get values() getter method instead of a direct static property.

akabiru committed:
Replace inheritance attempt with stimulus outlets instead

akabiru committed:
add failed contract error message

brunopagno committed:
remove individual workpackage permissions for comments with restricted visibility

akabiru committed:
Enhance wp note contract to include feature flag

akabiru committed:
[rubocop] Assignment Branch Condition size for wrapper_data_attributes is too high

Refactor via extract method proposed by @ulferts
https://github.com/opf/openproject/pull/18091#discussion_r1993284094

Retain singular method, breaking it apart doesn't add much value IMO

akabiru committed:
remove editorEscape event listener on disconnect

akabiru committed:
remove debugLog, seemingly unneeded

akabiru committed:
translation should be in base yml

akabiru committed:
Merge pull request #18262 from opf/implementation/62138-adjust-activity-tab-comment-dismiss-strategy-to-include-cancel-button-and-esc-keyboard-action

Implementation/62138 adjust activity tab comment editor dismiss strategy to be less eager, react to keyboard `Esc`

akabiru committed:
Merge pull request #18302 from opf/implementation/62139-implement-redesign-of-restricted-visibility-checkbox-placement-in-desktop-and-mobile

Implementation/62139 implement redesign of restricted visibility checkbox placement in desktop and mobile

akabiru committed:
Merge branch 'dev' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

brunopagno committed:
Merge pull request #18332 from opf/fix/60978-permissions-only-on-project

Remove specific work package permissions for comments with restricted visibility

akabiru committed:
stop event bubbling on kbd esc emit

akabiru committed:
[eslint] ensure values in comparison are of the same type

Functionality remains the same- just more explicit types to make typescript happy

akabiru committed:
Fix typo in comments

akabiru committed:
Allow browser confirm dialog

akabiru committed:
Enable comments ID URL feature in pull preview

akabiru committed:
Use method call instead

akabiru committed:
Define restricted extension on work package journals for query safety

  AR association extensions ensure the `restricted_visible` scope is only visible on
  the work package journals association.

  https://github.com/opf/openproject/pull/18091#discussion_r1993034943

akabiru committed:
Ensure cancel and submit buttons are always aligned to the right

brunopagno committed:
adjust query for restricted visibility "events"

akabiru committed:
Fix failing test in mentions

Ensure editor is closed before continuing assertions as we now require
explicit close actions.

---

Co-authored-by: Bruno Pagno <b.pagno@openproject.com>

brunopagno committed:
Merge branch 'implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes' into fix/60989-improve-fetcher-query

akabiru committed:
Replace implicit block with named parameter for clarity

https://github.com/opf/openproject/pull/18366/files#r2003254476

brunopagno committed:
adjust pluck to select + inline

akabiru committed:
Merge pull request #18366 from opf/implementation/60980-use-ar-extensions-for-safety

implementation/60980 - Define restricted extension on work package journals for query safety

brunopagno committed:
Merge pull request #18367 from opf/fix/60989-improve-fetcher-query

Improve fetcher query

brunopagno commented:
@akabiru in which scenario do we do squash&merge?

brunopagno commented:
@ulferts I think it's possible to review again. We should have covered the requested changes.

brunopagno commented:
Just noticed there are conflicts, I'll take a look

akabiru commented:
> @akabiru in which scenario do we do squash&merge?

I would merge in this case @brunopagno - I know it's a lot of commits but in this case I'd rather keep the history.

brunopagno committed:
Merge remote-tracking branch 'origin/dev' into implementation/60980-introduce-work-package-restricted-comments-as-part-of-journal-notes

brunopagno commented:
We addressed the concerns on the PR so we are merging it.

@ulferts if you want to have another look and point out anything else that we might have missed, feel free to reach out.
